File: RR.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <queue>

using namespace std;

// Function to calculate Turnaround Time and Waiting Time using Round Robin Scheduling
void roundRobin(vector<int> processes, vector<int> burstTime, vector<int> arrivalTime, int timeQuantum) {
    int n = processes.size();
    vector<int> remainingTime = burstTime;
    vector<int> completionTime(n), tat(n), wt(n);
    queue<int> readyQueue;

    int currentTime = 0;
    int completed = 0;
    vector<bool> visited(n, false);

    // Add processes to ready queue as per their arrival time
    for (int i = 0; i < n; i++) {
        if (arrivalTime[i] <= currentTime && !visited[i]) {
            readyQueue.push(i);
            visited[i] = true;
        }
    }

    while (!readyQueue.empty() || completed < n) {
        if (!readyQueue.empty()) {
            int currentProcess = readyQueue.front();
            readyQueue.pop();

            if (remainingTime[currentProcess] > 0) {
                int timeToRun = min(remainingTime[currentProcess], timeQuantum);
                currentTime += timeToRun;
                remainingTime[currentProcess] -= timeToRun;

                // If the process is completed
                if (remainingTime[currentProcess] == 0) {
                    completionTime[currentProcess] = currentTime;
                    tat[currentProcess] = completionTime[currentProcess] - arrivalTime[currentProcess];
                    wt[currentProcess] = tat[currentProcess] - burstTime[currentProcess];
                    completed++;
                }
            }

            // Add newly arrived processes to the queue
            for (int i = 0; i < n; i++) {
                if (arrivalTime[i] <= currentTime && !visited[i]) {
                    readyQueue.push(i);
                    visited[i] = true;
                }
            }

            // If the process is not completed, add it back to the ready queue
            if (remainingTime[currentProcess] > 0) {
                readyQueue.push(currentProcess);
            }
        } else {
            currentTime++; // No process is ready, move time forward
            for (int i = 0; i < n; i++) {
                if (arrivalTime[i] <= currentTime && !visited[i]) {
                    readyQueue.push(i);
                    visited[i] = true;
                }
            }
        }
    }

    // Output results
    cout << "\nRound Robin Scheduling:\n";
    cout << "Process\tArrival Time\tBurst Time\tTurnaround Time\tWaiting Time\n";
    double avgTAT = 0, avgWT = 0;

    for (int i = 0; i < n; i++) {
        cout << "P" << processes[i] << "\t\t" << arrivalTime[i] << "\t\t" << burstTime[i]
             << "\t\t" << tat[i] << "\t\t" << wt[i] << '\n';
        avgTAT += tat[i];
        avgWT += wt[i];
    }

    cout << "\nAverage Turnaround Time: " << fixed << setprecision(2) << avgTAT / n << endl;
    cout << "Average Waiting Time: " << fixed << setprecision(2) << avgWT / n << endl;
}

// Main function
int main() {
    int n, timeQuantum;

    cout << "Enter the number of processes: ";
    cin >> n;

    vector<int> processes(n), burstTime(n), arrivalTime(n);

    cout << "Enter Arrival Time and Burst Time for each process:\n";
    for (int i = 0; i < n; i++) {
        cout << "Process " << i + 1 << ":\n";
        cout << "Arrival Time: ";
        cin >> arrivalTime[i];
        cout << "Burst Time: ";
        cin >> burstTime[i];
        processes[i] = i + 1;
    }

    cout << "Enter Time Quantum: ";
    cin >> timeQuantum;

    roundRobin(processes, burstTime, arrivalTime, timeQuantum);

    return 0;
}

File: RR.py
def round_robin(processes, burst_time, arrival_time, time_quantum):
    n = len(processes)
    remaining_time = burst_time[:]
    completion_time = [0] * n
    waiting_time = [0] * n
    turnaround_time = [0] * n
    time = 0
    queue = []
    visited = [False] * n
    completed = 0

    while completed < n:
        # Add processes to the queue based on arrival time
        for i in range(n):
            if arrival_time[i] <= time and not visited[i]:
                queue.append(i)
                visited[i] = True
        
        if queue:
            # Get the first process from the queue
            current = queue.pop(0)
            
            # Execute the process
            if remaining_time[current] > time_quantum:
                time += time_quantum
                remaining_time[current] -= time_quantum
            else:
                time += remaining_time[current]
                completion_time[current] = time
                remaining_time[current] = 0
                completed += 1
            
            # Add the process back to the queue if not completed
            if remaining_time[current] > 0:
                queue.append(current)
            
            # Add newly arrived processes to the queue
            for i in range(n):
                if arrival_time[i] <= time and not visited[i]:
                    queue.append(i)
                    visited[i] = True
        else:
            time += 1  # If no process is ready, increment time

    # Calculate TAT and WT
    for i in range(n):
        turnaround_time[i] = completion_time[i] - arrival_time[i]
        waiting_time[i] = turnaround_time[i] - burst_time[i]

    # Display the results
    print(f"Round Robin Scheduling (Time Quantum = {time_quantum}) Results:")
    print(f"{'Process':<10}{'Arrival Time':<15}{'Burst Time':<15}{'Completion Time':<20}{'TAT':<10}{'WT':<10}")
    for i in range(n):
        print(f"{processes[i]:<10}{arrival_time[i]:<15}{burst_time[i]:<15}{completion_time[i]:<20}{turnaround_time[i]:<10}{waiting_time[i]:<10}")

    # Calculate and display averages
    avg_tat = sum(turnaround_time) / n
    avg_wt = sum(waiting_time) / n
    print(f"\nAverage Turnaround Time: {avg_tat:.2f}")
    print(f"Average Waiting Time: {avg_wt:.2f}")


# Example Input
processes = ["P1", "P2", "P3", "P4"]
burst_time = [5, 8, 7, 3]
arrival_time = [0, 1, 2, 3]
time_quantum = 3

# Run Round Robin Scheduling
print("===" * 10)
round_robin(processes, burst_time, arrival_time, time_quantum)
